pkg/ipam-node/allocator/allocator.go

```diff
@@ -0,0 +1,181 @@
+}
+
+type allocator struct {
+	rangeSet *RangeSet
+	store    storePkg.Store
+	poolName string
+}
+
+func NewIPAllocator(s *RangeSet, poolName string, store storePkg.Store) IPAllocator {
+	return &allocator{
+		rangeSet: s,
+		store:    store,
+		poolName: poolName,
+	}
+}
+
+// Allocate allocates an IP
+func (a *allocator) Allocate(id string, ifName string, meta types.ReservationMetadata) (*current.IPConfig, error) {
+	var reservedIP *net.IPNet
+	var gw net.IP
+
+	iter := a.getIter()
+	for {
+		reservedIP, gw = iter.Next()
+		if reservedIP == nil {
+			break
+		}
+
+		err := a.store.Reserve(a.poolName, id, ifName, meta, reservedIP.IP)
+		if err == nil {
+			break
+		}
+		if !errors.Is(err, storePkg.ErrIPAlreadyReserved) {
+			return nil, err
+		}
+	}
+
+	if reservedIP == nil {
+		return nil, ErrNoFreeAddresses
+	}
+
+	return &current.IPConfig{
+		Address: *reservedIP,
+		Gateway: gw,
+	}, nil
+}
+
+type RangeIter struct {
+	rangeSet *RangeSet
+	// The current range id
+	rangeIdx int
+	// Our current position
+	cur net.IP
+	// The IP where we started iterating; if we hit this again, we're done.
+	startIP net.IP
+}
+
+// getIter encapsulates the strategy for this allocator.
+// We use a round-robin strategy, attempting to evenly use the whole set.
+// More specifically, a crash-looping container will not see the same IP until
+// the entire range has been run through.
+// We may wish to consider avoiding recently-released IPs in the future.
+func (a *allocator) getIter() *RangeIter {
+	iter := RangeIter{
+		rangeSet: a.rangeSet,
+	}
+
+	// Round-robin by trying to allocate from the last reserved IP + 1
+	startFromLastReservedIP := false
+
+	// We might get a last reserved IP that is wrong if the range indexes changed.
+	// This is not critical, we just lose round-robin this one time.
+	lastReservedIP := a.store.GetLastReservedIP(a.poolName)
+	if lastReservedIP != nil {
+		startFromLastReservedIP = a.rangeSet.Contains(lastReservedIP)
+	}
+
+	// Find the range in the set with this IP
+	if startFromLastReservedIP {
+		for i, r := range *a.rangeSet {
+			if r.Contains(lastReservedIP) {
+				iter.rangeIdx = i
+
+				// We advance the cursor on every Next(), so the first call
+				// to next() will return lastReservedIP + 1
+				iter.cur = lastReservedIP
+				break
+			}
+		}
+	} else {
+		iter.rangeIdx = 0
+		iter.startIP = (*a.rangeSet)[0].RangeStart
+	}
+	return &iter
+}
+
+// Next returns the next IP, its mask, and its gateway. Returns nil
+// if the iterator has been exhausted
+func (i *RangeIter) Next() (*net.IPNet, net.IP) {
+	r := (*i.rangeSet)[i.rangeIdx]
+
+	// If this is the first time iterating, and we're not starting in the middle
+	// of the range, then start at rangeStart, which is inclusive
+	if i.cur == nil {
+		i.cur = r.RangeStart
+		i.startIP = i.cur
+		if i.cur.Equal(r.Gateway) {
+			return i.Next()
+		}
+		return &net.IPNet{IP: i.cur, Mask: r.Subnet.Mask}, r.Gateway
+	}
+
+	nextIP := ip.NextIP(i.cur)
```

Comment:
is there really a case where nextIP is nil ?

Reply:
nextIP can be nil in case of IP overflow. This should not happen inside this package, but I would still prefer to check for nil here